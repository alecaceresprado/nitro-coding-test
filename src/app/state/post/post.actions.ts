import { createAction, props } from '@ngrx/store';

import { Post } from '@models';

export const fetchPosts = createAction('[POST] fetch posts');
export const fetchPostsSucceeded = createAction(
  '[POST] fetch post - SUCCESS',
  props<{ posts: Post[] }>()
);
export const fetchPostsFailed = createAction('[POST] fetch post - FAIL');
export const updatePost = createAction(
  '[POST] update post',
  props<{ id: number; location: string; author: string }>()
);
