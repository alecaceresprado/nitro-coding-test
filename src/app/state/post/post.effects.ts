import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { switchMap, map, catchError } from 'rxjs/operators';

import { PostService } from '@services';
import {
  fetchPosts,
  fetchPostsSucceeded,
  fetchPostsFailed,
} from './post.actions';
import { Post } from '@models';

@Injectable()
export class PostEffects {
  /*
   * @listen [POST] fetch posts
   * @dispatch [POST] fetch post - SUCCESS
   * @dispatch [POST] fetch post - FAIL
   */
  public fetchPosts$ = createEffect(() =>
    this.actions$.pipe(
      ofType(fetchPosts),
      switchMap(() =>
        this.postService.getPosts().pipe(
          map((posts: Post[]) => fetchPostsSucceeded({ posts })),
          catchError(() => of(fetchPostsFailed()))
        )
      )
    )
  );

  constructor(private actions$: Actions, private postService: PostService) {}
}
