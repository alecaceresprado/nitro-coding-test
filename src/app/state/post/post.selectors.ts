import { createFeatureSelector } from '@ngrx/store';

import { Post } from '@models';

export const getPosts = createFeatureSelector<Post[]>(
  'posts'
);
