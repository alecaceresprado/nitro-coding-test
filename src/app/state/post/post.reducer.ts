import { createReducer, Action, on } from '@ngrx/store';

import { fetchPostsSucceeded, updatePost } from './post.actions';
import { Post } from '@models';

export const postInitialState: Post[] = [];

const reducer = createReducer(
  postInitialState,
  on(fetchPostsSucceeded, (state, { posts }) => posts),
  on(updatePost, (state, {id, author, location}) =>
    state.map(post => {
      if(post.id === id) {
        return {
          ...post,
          location,
          author
        };
      }
      return post;
    })
  )
);

export function postReducer(state: Post[] | undefined, action: Action): Post[] {
  return reducer(state, action);
}
