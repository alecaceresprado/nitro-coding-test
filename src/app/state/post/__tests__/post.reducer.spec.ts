import { postReducer } from '../post.reducer';
import { fetchPostsSucceeded, updatePost } from '../post.actions';

const post1 = {
  id: 1,
  location: 'San Francisco',
  time: '1552657573',
  author: 'Happy User',
  text:
    'Proper PDF conversion ensures that every element of your document remains just as you left it.',
};

const post2 = {
  id: 3,
  location: 'San Francisco',
  time: '1501690872999',
  author: 'Happy Developer',
  text: 'Digital transformation isnâ€™t justa buzzword',
};

describe('postReducer', () => {
  it('should return posts from store', () => {
    const result = postReducer([], fetchPostsSucceeded({ posts: [post1] }));
    expect(result).toEqual([post1]);
  });

  it('should edit post', () => {
    const result = postReducer(
      [post1, post2],
      updatePost({ id: 1, location: 'Sydney', author: 'Happy Candidate' })
    );
    expect(result).toMatchSnapshot();
  });
});
