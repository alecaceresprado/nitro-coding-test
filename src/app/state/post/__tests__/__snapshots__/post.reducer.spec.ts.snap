// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postReducer should edit post 1`] = `
Array [
  Object {
    "author": "Happy Candidate",
    "id": 1,
    "location": "Sydney",
    "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
    "time": "1552657573",
  },
  Object {
    "author": "Happy Developer",
    "id": 3,
    "location": "San Francisco",
    "text": "Digital transformation isnâ€™t justa buzzword",
    "time": "1501690872999",
  },
]
`;
